### PR TITLE
feat(core): Security Headers for Pre-rendered Routes

### LINKED ISSUE #386: [Feature Request]: expose generated security headers at build time
### Is your feature request related to a problem? Please describe.
I use SSG and I want to be able to extract the generated security headers (especially the csp with defined integrity hashes), so that I can write those headers to an nginx configuration file before deploying the static build output to my nginx server.

### Describe the solution you'd like
I'm not sure how this feature can be implemented best, but here are some vague ideas I have:

- Fix [05-cspSsgPresets.ts](https://github.com/Baroshem/nuxt-security/blob/main/src/runtime/nitro/plugins/05-cspSsgPresets.ts) somehow
- Write the headers to a file so a user can read the file and use it's content
- Provide a hook that exposes the headers once they are generated for a user to hook into (similar to the nitro compiled hook)

### Describe alternatives you've considered
I tried to use the runtime hooks feature introduced in v1.1.0, but this didn't work.

Another solution for me would be to calculate the integrity hashes myself in some script and then use those values to create my header configuration, but it makes more sense to use this module, because it also forces the header values in dev mode, so they can be tested and it gives some common sense defaults that I wouldn't have to configure manually.

### Additional context
Related to #336.

### LINKED ISSUE #434: Support for Security Headers for resources other than HTML
Hello,

I've noticed that all web security headers only appear in the HTML responses, but not in the Nuxt resources such as CSS, JS, etc. Public resources are also not considered by Nuxt-Security.

Security scanners are raising alarms because the necessary headers, such as X-Content-Type-Options: nosniff, are not set on the resources.

Currently, I've solved this issue with an additional proxy that sets these headers on all resources. However, I suspect this might be a bug in Nuxt-Security, but I'm not sure as I'm not very familiar with this area.

Is it intended that the security headers are only set in the HTML responses? Or is this a bug?

Thank you in advance for your help.

Best regards

### PR DESCRIPTION
**Pull request by vejja**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [x] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Description

With this PR, security headers
- are now available in Hybrid Applications (see #409)
- are also available in Client-Side Rendered Applications via Nitro presets or hook (see #386)
- are available on non-HTML resources, where relevant (see #434)

Previously, when a route was pre-rendered via `routeRules`, the Nuxt application could not deliver the Security Headers. This is because the HTML page is served directly as a pre-built static asset, and we could not intercept the headers in that case to modify them.

This PR proposes a modification of the core engine of Nuxt Security, whereby we now record all pre-rendered pages at build time, and then intercept all requests to verify whether they correspond to a pre-rendered page.

As a consequence of this core upgrade, we now deliver five additional features:
1. For pre-rendered pages in a Nuxt Hybrid application, we are now able to provide **CSP by way of headers**, in addition to the `meta` tag. This addresses the issue where some security scanners are complaining about CSP compliance because they do not check the `meta` tag.
2. More generally, we are now able to provide _all_ security headers for pre-rendered pages. Previously, only CSP could be provided through the `meta` tag, but now **all other policies** (e.g. COEP, STS, etc...) can also be delivered via headers for pre-rendered pages in Hybrid Mode.
3. We take this opportunity to also re-activate the `cspSsgPresets` feature, which had to be deactivated after the upgrade to Nuxt 3.9.3 caused issue #348. With this feature, we now generate a `headers.json` file that is used by Nitro in static presets. For instance, a **static website** deployed on Vercel or Netlify now has security headers natively on all pages, even without a server runtime !
4. We introduce a new `nuxt-security:prerenderedHeaders` build-time hook, which allows to further extend the previous feature **when there is no Nitro preset available**. This hook can be used to manually generate a headers configuration file for any custom static deployment.
5. In addition, we now add some common security headers on **non-HTML assets**, such as images, `.js` or `.css` files, api requests, etc. The following headers are now delivered on all resources: `referrerPolicy`, `strictTransportSecurity`, `xContentTypeOptions`, `xDownloadOptions`, `xFrameOptions`, `xPermittedCrossDomainPolicies`, and `xXSSProtection`.

Please note that : 
- The features 1. and 2. are available under a new `ssg: nitroHeaders` option (default `true`). This option is modifiable both globally and at route-level.
- Feature 3. is available under a new `ssg: exportToPresets` option (default `true`). This option is only available globally.
- Features 4. and 5. are always available (no option switch).

Closes #386 
Closes #409 
Closes #434

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes (if not applicable, please state why)

### CODE REVIEW

**Comment by Baroshem** [comment]
Hi @vejja 

Thanks for this amazing PR!

I needed to first release your other PR with SRI fix and it caused the conflicts in this PR. Could you please rebase it? :)

Also, regarding the release of this feature, I would aim for 1.5.0 next week probably if that is ok with you? :)

I will also review it in the upcoming days to give you some feedback.

**Comment by vejja** [comment]
I'm also closing #434 with this PR
Will update description and notes

**Comment by vejja** [comment]
@Baroshem 
As a side technical note, I am also fixing the issue that popped up with #438, where we had to duplicate some `utils` in different folders because runtime code could not access files in the `src` folder.
I am adding a `build.config.ts` file that tells `unjs/unbuild` to make the `utils` code available at the root. 
This allows to put all common code in `utils`, no duplication required anymore !

File: build.config.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { defineBuildConfig } from "unbuild"
+export default defineBuildConfig({
+  declaration: true,
+  entries: [
+    'src/utils/hash.ts',
+    'src/utils/headers.ts',
+    'src/utils/merge.ts'
+
```
Comment (Baroshem) [inline build.config.ts:8]: nitpick: unnecessary break line

File: playground/nuxt.config.ts [hunk 2]
```diff
@@ -55,6 +56,24 @@ export default defineNuxtConfig({
       interval: 30000,
       headers: true
     },
-    removeLoggers: false
+    removeLoggers: false,
+  },
+
+  hooks: {
+    'nuxt-security:prerenderedHeaders': async(prerenderedHeaders) => {
+      console.log('in prerenderedHeaders hook', prerenderedHeaders)
```
Comment (Baroshem) [inline playground/nuxt.config.ts:64]: question: do we want this console.log to be here?

**Review by Baroshem — APPROVED** [review]
Two small nitpicks, overall really great work!

**Comment by Baroshem** [comment]
Amazing work @vejja !

I will merge it to the 2.0.0-rc.1 branch where we can do some additional testing and prepare a big release soon :)

**Pull request merged**